### PR TITLE
Add integration tests for VM snapshot support

### PR DESCRIPTION
**Pull request by connorkuehl**
Closes: https://github.com/harvester/tests/issues/363

There's an additional commit that adds an (incomplete) VolumeSnapshotManager. This is because one of the test cases asserts that the volume snapshot is automatically cleaned up after the volume it relates to has been deleted, so the test case asserts that it exists, then after the volume is deleted, it is used to assert that it no longer exists. I can split this commit out to a separate PR to implement it more completely, if that is preferred.

### CODE REVIEW

**Comment by connorkuehl** [comment]
I was originally waiting to plug this PR into the automated testing infrastructure but I'm still trying to get access to that, but it's taking a long time to get access to that infrastructure. I think this is reviewable in the meantime and the tests run to completion on my local machine anyway, so.

**Comment by connorkuehl** [comment]
Latest revision removes the redundant backup test cases as requested :slightly_smiling_face:

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+
+
+@pytest.fixture(scope="class")
+def backup_source_vm(newvm_name, api_client, ssh_keypair, image, wait_timeout):
+    yield create_vm(newvm_name, api_client, ssh_keypair, image, wait_timeout)
+    delete_vm(newvm_name, api_client, wait_timeout)
+
+
+@pytest.fixture(scope="class")
+def vm_backup(vm_backup_name, backup_source_vm,
```
Comment (khushboo-rancher) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:229]: Are these backup related fixtures required?

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+                else:
+                    break
+            else:
+                raise AssertionError(f"Unable to login to {name}")
+
+            with vm_sh as sh:
+                action(sh)
+
+
+@pytest.mark.p1
```
Comment (khushboo-rancher) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:353]: Can we mark this P0 test case?

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+        3. A file called `test.txt` containing data: `123`.
+        4. A snapshot called `vm-snapshot`.
+
+        Since all of the heavy lifting is done in the fixture,
+        assert the snapshot was created successfully for this test.
+        """
+        code, data = api_client.vm_snapshots.get(vm_snapshot)
+
+        assert 200 == code
+        assert data.get("status", {}).get("readyToUse") is True
```
Comment (khushboo-rancher) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:379]: These checks are already there in the fixture. :) 
Instead of having the logic of creating snapshot and checking the status etc, can we move those steps in the test. We can use the fixtures to create VM etc. The reason for the asked changes are as below:
1. By having the steps in the test will make it more readable.
2. We will isolate the test. As fixture can be used at many places and making a change to only this test will become difficult in future.

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]
Thanks @connorkuehl for the PR.

I feel we can make some changes to make test cases isolated and more readable by having the logic of test steps in the test instead of writing them in fixtures. We should definitely use fixture to write the prerequisite of tests and cleaning up them. I've tried to explain my point in one of the below comment. Let me know we can sync up to discuss this.

File: harvester_e2e_tests/integration/test_vm_snapshot.py (hunk 1 above)
Comment (connorkuehl) [reply harvester_e2e_tests/integration/test_vm_snapshot.py]: No, I've removed them from the latest revision.

File: apiclient/harvester_api/managers.py [hunk 4]
```diff
@@ -221,6 +221,14 @@ def export(self, name, image_name, storage_class, namespace=DEFAULT_NAMESPACE, *
         return self._create(path, params=params, json=export_spec, raw=raw)
 
 
+class VolumeSnapshotManager(BaseManager):
```
[the remark this answers is not in the document]
Comment (connorkuehl) [reply apiclient/harvester_api/managers.py:224]: The latest revision has `delete`, but implementing the spec for VolumeSnapshot will take a little more time. I've pushed the latest revision so we can get the ball rolling on reviewing the rest of the PR in the meantime.

File: harvester_e2e_tests/integration/test_vm_snapshot.py (hunk 2 above)
Comment (connorkuehl) [reply harvester_e2e_tests/integration/test_vm_snapshot.py]: Updated this in the latest revision

File: harvester_e2e_tests/integration/test_vm_snapshot.py (hunk 3 above)
Comment (connorkuehl) [reply harvester_e2e_tests/integration/test_vm_snapshot.py:274]: Incorporated this in the latest revision

**Comment by connorkuehl** [comment]
@khushboo-rancher Thanks for the review! 🙂  I've incorporated most of the feedback with the exception of the create method for the VolumeSnapshotManager (see response in other thread), I'll keep working on it, but hopefully the rest of the PR is ready for another pass

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
+from datetime import datetime, timedelta
+from time import sleep
+from paramiko.ssh_exception import ChannelException
+
+
+import pytest
+import yaml
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:7]: based on [PEP8](https://peps.python.org/pep-0008/#imports)
import should be grouped as:
```python
import yaml
from datetime import datetime, timedelta
from time import sleep

from paramiko.ssh_exception import ChannelException
import pytest
```

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+        machine results in a virtual machine with the expected
+        well-known file (`test.txt`) with the expected contents
+        (`123`).
+
+        Prerequisites:
+        1. The source VM from the first test case and its
+           snapshot (`vm-snapshot`).
+        2. A VM restored from `vm-snapshot`.
+        """
+
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:292]: I think this implement will not hit criteria, the VM created and took snapshot, but we didn't modify it after snapshot, so we can not guarantee the VM really restored from the snapshot or not.

and while restoring to new VM, we will not stop the source VM.

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+                                         ssh_keypair, host_shell,
+                                         vm_shell, wait_timeout):
+        name, ssh_user = source_vm
+
+        stop_vm(name, api_client, wait_timeout)
+
+        """
+        Test that the Harvester API rejects a `replace`
+        VirtualMachineRestore where the deletePolicy is
+        not `retain`.
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:316]: we should separate this into another test case.

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+        assert wantmsg in reason
+        assert 422 == code
+
+        """
+        Test that the original virtual machine can be replaced
+        from its original snapshot (`vm-snapshot`) and that
+        the snapshot's data contains the well-known file (`test.txt`)
+        and its expected contents (`123`).
+
+        Prerequisites:
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:339]: same as previous `restore to new`, we didn't modify the VM after snapshot, so we can't guarantee the VM really restored from the snapshot or not.

**Review by lanfon72 — CHANGES_REQUESTED** [review]

File: harvester_e2e_tests/integration/test_vm_snapshot.py (hunk 6 above)
Comment (connorkuehl) [reply harvester_e2e_tests/integration/test_vm_snapshot.py:304]: 🤔 I understand this logic for the test_replace_vm_with_vm_snapshot, but I'm not sure I follow for this test case. Should I add more assertions to the test proving that this is accessing a completely separate/new VM?

Comment (lanfon72) [reply harvester_e2e_tests/integration/test_vm_snapshot.py:304]: In manual testing, we will test this as following steps:
1. Create the VM `vm1`
2. Write data into `vm1` (`vm1-state-1`)
3. Take snapshot for `vm1`
4. Write or modify exisiting data into `vm1` (`vm1-state-2`)
5. Restore snapshot into `vm1` or `vm1-new`
6. Data in restored VM should keep in `vm1-state1`, not `vm1-state-2`

**Comment by khushboo-rancher** [comment]
@lanfon72 This is ready for review, could you please review it whenever get a chance?

File: harvester_e2e_tests/integration/test_vm_snapshot.py [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+        `vm-snapshot` VM snapshot exists.
+        """
+        name, ssh_user = source_vm
+
+        stop_vm(name, api_client, wait_timeout)
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_vm_snapshot.py:380]: I think the implementation is still not fit the test criteria, data in the `source_vm` is the same as the snapshot taken, so we can't say that the restore is success or not.

what we test a snapshot is like:
1. Create `vm1`
2. Write data into `vm1`, naming its state as `a`
3. Take snapshot `vm1-s` for `vm1`, so the snapshot is in same state `a` of the `vm1`
4. Write or delete data into `vm1`, naming its state as `b`
5. Restore snapshot `vm1-s` for `vm1`
6. `vm1` should back to state `a`

Steps 4 is the criteria for the test case, then we can guarantee that restore the snapshot is succeed because the state is changed.

For this implementation, Step 4 is missing, so we can't guarantee that the snapshot is restored or not.

**Review by lanfon72 — CHANGES_REQUESTED** [review]
All related test cases should change (write or delete) data in VM before restore, to make sure the restore will recover into previous state.

File: harvester_e2e_tests/integration/test_vm_snapshot.py (hunk 9 above)
Comment (lanfon72) [reply harvester_e2e_tests/integration/test_vm_snapshot.py:380]: We will not make the side effect in the test case `test_vm_snapshot_create`, the test case is purely require that create a snapshot, modify it after snapshot is not included.

Without the side effect, when we restore the VM into **new** one, we can simply compare files in both VMs to make sure it restored correctly.

And If we are going to verify restore existing VM, we would update the VM before restore to make sure what we changed will be reverted.

Based on this, we will not care about what we wrote into, so there will not have some constant assertion like `assert "123" in out`, because we will compare files in two VMs or check the modification is reverted.

Comment (connorkuehl) [reply harvester_e2e_tests/integration/test_vm_snapshot.py:380]: > We will not make the side effect in the test case test_vm_snapshot_create,

The code currently _does_ make this side-effect. Just to make sure I'm parsing your comment correctly, are you saying that you want me to remove that side-effect from the test case and then duplicate it across all of the other relevant test cases? I put it there initially because I didn't want every test case to have a connect, write, wait 2-seconds loop, which would extend the runtime of the test suite.

Comment (lanfon72) [reply harvester_e2e_tests/integration/test_vm_snapshot.py:380]: yes it is necessary.
When we are just execute `test_vm_snapshot_create`, what we will expects is, the VM having the same state as the snapshot been taken, modify the VM is not required and not expected.

For other test cases, we will not depends on the _side effect_ of data what we wrote, in **restore new**, we would compare two VMs having same files and content, not compare the constants from the `test_vm_snapshot_create`; in **restore existing**, we will just check what we wrote/deleted before restore, will be reverted, not compare the constants from the `test_vm_snapshot_create`.

In current implementation, we would need make the filename comes from `source_vm`, then the implementation would be more stable to extend when we are going to change the data what we are going to wrote, we would just need to store file names and passing them in test cases.

Comment (connorkuehl) [reply harvester_e2e_tests/integration/test_vm_snapshot.py:401]: Sorry for the delay! 🙂 

I think I've addressed this in the latest revision. The new test fixtures will overwrite contents prior to restore and assert that the restored or replaced VM have the original contents prior to the overwrite.

**Review by lanfon72 — APPROVED** [review]
LGTM, thx !

**Comment by khushboo-rancher** [comment]
Thanks @connorkuehl and @lanfon72.
@connorkuehl could you please fix the `GH-Pages` error before we merge this?
```
Error: testcases/virtualmachines/scheduling.spec.ts:17:46 - error TS7006: Parameter 'host' implicitly has an 'any' type.

1[7](https://github.com/harvester/tests/actions/runs/6090066516/job/16524173011?pr=817#step:9:8)     const hostNames: string[] = hostList.map(host => host.name);
```

**Review by khushboo-rancher — APPROVED** [review]
